feat(simplecache): allow specifying custom simplecache path

hypeJunction opened the pull request:
Sites can now define a custom location to store cached views by providing
$CONFIG->assetroot value.

hypeJunction committed:
feat(simplecache): allow specifying custom simplecache path

Sites can now define a custom location to store cached views by providing
$CONFIG->assetroot value.

hypeJunction committed:
chore(views): centralize cachable view registration in views service

jeabakker commented on engine/classes/Elgg/Di/ServiceProvider.php:
```diff
@@ -751,12 +751,19 @@ public function initConfig(Config $config, ServiceProvider $sp) {
+		if ($config->assetroot) {
+			$config->assetroot = Paths::sanitize($config->assetroot);
+		} else {
+			$config->assetroot = Paths::sanitize('views_simplecache');
```
shouldn't this be something like `$config->cachroot . 'views_simplecache'`?

jeabakker replied:
I mean in `Paths::sanitize`

hypeJunction replied:
Yes, it should. Question for you, do we want to keep these BC? If someone has a symlink setup for the old location, we would probably want to keep it there, no?

jeabakker replied:
Or rewrite the symlink to the new location (if possible). If not possible maybe add admin notice.

I think don't try to hard to keep BC, just mention something should manually be done

hypeJunction committed:
chore(config): simplify config initialization and validation

Path values in Elgg config are now sanitized/normalized on __set().
Helper functions no longer provide default values for paths,
instead they are set when service provider is initialized
(this helps avoid discrepancies in values returned by helper
functions and those set in config).

hypeJunction committed:
fix(filesystem): do not rely on sys_get_temp_dir() to be writable

In cases when system temp directory is not writable, temp
files are written to tmp/ in project root.
File system caches will no longer be written unless cachepath
is set explicitly in config, i.e. during installation before
config is fully instantiated.

hypeJunction committed:
chore(tests): clean up upgrades before test

hypeJunction commented:
To hell with it. Another endless snowball.
